R/shiny.R

```diff
@@ -58,11 +64,25 @@ renderDataTable = function(expr, env = parent.frame(), quoted = FALSE, ...) {
     instance = exprFunc()
     if (!all(c('datatables', 'htmlwidget') %in% class(instance))) {
       instance = datatable(instance, ...)
+    } else {
+      if (length(list(...)) != 0) {
+        warning("renderDataTable ignores ... arguments when expr yields a datatable object; see ?renderDataTable")
+      }
     }
 
-    if (is.function(hook <- instance$preRenderHook))
-      instance = hook(instance, currentSession, currentOutputName)
-    instance$preRenderHook = NULL
+    # in the server mode, we should not store the full data in JSON
+    if (server) {
+      origData = instance[['x']][['data']]
+      instance$x$data = NULL
```

Comment:
@yihui recently reminded me that using `instance$x$data` for reads will do partial matching, while `[[` will not. For writes `$` is fine.

I don't know that I'll do `[[` style reads forever, but just today I could not stomach the partial matching... :-\